packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js

```diff
@@ -4797,6 +4797,115 @@ QUnit.module('the \'acceptCustomValue\' option', moduleSetup, () => {
+                acceptCustomValue: true,
+                onCustomItemCreating(args) {
+                    const newValue = args.text;
+                    const { component } = args;
+                    const currentItems = component.option('items');
+                    const isItemInDataSource = currentItems.some((item) => item === newValue);
+                    if(!isItemInDataSource) {
+                        currentItems.unshift(newValue);
+                        component.option('items', currentItems);
+                    }
+                    args.customItem = newValue;
+                },
+                deferRendering
+            });
+            const instance = $element.dxTagBox('instance');
+            const $input = $element.find(`.${TEXTEDITOR_INPUT_CLASS}`);
+            const keyboard = keyboardMock($input, true);
+            keyboard
+                .focus()
+                .type(firstCustomValue)
+                .press('enter')
+                .type(secondCustomValue)
+                .press('enter');
+
+
+            if(deferRendering) {
+                const isValueExist = [firstCustomValue, secondCustomValue].every(value =>
+                    instance.option('selectedItems').includes(value)
+                );
+                assert.ok(isValueExist, 'Custom Value is properly selected');
+            } else {
+                const isValueExist = instance.option('selectedItem').includes(secondCustomValue);
+                assert.ok(isValueExist, 'All custom values is properly selected');
+            }
+        });
+
+        QUnit.test(`if ${deferRendering} should include customValue in list of items [T1230995]`, function(assert) {
+            const customValue = 'test 1';
+            const $element = $('#tagBox').dxTagBox({
+                applyValueMode: 'useButtons',
+                items: ['Item 1', 'Item 2', 'Item 3'],
+                acceptCustomValue: true,
+                onCustomItemCreating(args) {
+                    const newValue = args.text;
+                    const { component } = args;
+                    const currentItems = component.option('items');
+                    const isItemInDataSource = currentItems.some((item) => item === newValue);
+                    if(!isItemInDataSource) {
+                        currentItems.unshift(newValue);
+                        component.option('items', currentItems);
+                    }
+                    args.customItem = newValue;
+                },
+                deferRendering
+            });
+            const instance = $element.dxTagBox('instance');
+            const $input = $element.find(`.${TEXTEDITOR_INPUT_CLASS}`);
+            const keyboard = keyboardMock($input, true);
+            keyboard
+                .focus()
+                .type(customValue)
+                .press('enter');
+
+            const isValueExist = instance.option('items').includes(customValue);
```

Comment:
It's better to check the actual values of items rather than the presence of a new item in the list.
```
assert.strictEqual(this.option('items'), expectedItems, 'Custom value is included in tagbox items');
```

Reply:
[30d85eb](https://github.com/DevExpress/DevExtreme/pull/27341/commits/30d85eb8b885fb8cfad6063c4b073a2970efa847)